storage/queue.go

```diff
@@ -86,40 +94,42 @@ type baseQueue struct {
 // maxSize doesn't prevent new ranges from being added, it just
 // limits the total size. Higher priority ranges can still be
 // added; their addition simply removes the lowest priority range.
-func newBaseQueue(shouldQ shouldQueueFn, maxSize int) *baseQueue {
+func newBaseQueue(shouldQ shouldQueueFn, process processFn, maxSize int) *baseQueue {
 	return &baseQueue{
 		shouldQ: shouldQ,
+		process: process,
 		maxSize: maxSize,
 		ranges:  map[int64]*rangeItem{},
 	}
 }
 
-// length returns the current size of the queue.
-func (bq *baseQueue) length() int {
+// Length returns the current size of the queue.
+func (bq *baseQueue) Length() int {
 	return bq.priorityQ.Len()
 }
 
-// next dequeues and returns the highest priority range. If the queue
-// is empty, returns nil.
-func (bq *baseQueue) next() *Range {
+// Pop dequeues and processes the highest priority range in the queue.
+// Returns the range if not empty; otherwise, returns nil.
+func (bq *baseQueue) Pop() *Range {
 	if bq.priorityQ.Len() == 0 {
 		return nil
 	}
 	item := heap.Pop(&bq.priorityQ).(*rangeItem)
 	delete(bq.ranges, item.value.Desc.RaftID)
+	bq.process(time.Now(), item.value)
```

Comment:
I'm not quite sure yet about how we're going to run these. Might be a thread of operation per queue, might be one thread for all queues and prioritize the queues. Might be some number of active workers > 1. We'll see how it shakes out in practice. This part is not really operational, mostly just placeholder.

Fixed logging issue.